### PR TITLE
Go sharded client

### PR DESCRIPTION
**Pull request by kingster**
Implements the sharded go-client #40

### CODE REVIEW

File: clients/go/dkv/client.go [hunk 1]
```diff
@@ -0,0 +1,248 @@
...
+	h := xxhash.New64()
+	h.Write(key)
+	hash := h.Sum64()
+	var id = (hash & 0xFFFF) % uint64(len(p.ShardConfiguration))
+	//LongHashFunction xx = LongHashFunction.xx();
+	//long hsh = xx.hashBytes(key);
+	//return (int) ((hsh & 0xFFFF) % shardConfiguration.getNumShards());
+	return int(id)
+}
+func (p KeyHashBasedShardProvider) ProvideShard(key []byte) (*DKVShard, error) {
```
Comment (KalyanAkella) [inline clients/go/dkv/client.go:71]: Modify the type of the receiver `p` to `*KeyHashBasedShardProvider`. Prefer pointer receivers for structs.

Comment (KalyanAkella) [inline clients/go/dkv/client.go:71]: Please include GoDoc (along with a new line before line 71) for this public method.

File: clients/go/dkv/client.go [hunk 2]
```diff
@@ -0,0 +1,248 @@
...
+	shardId := p.getShardID(key)
+	return &p.ShardConfiguration[shardId], nil
+}
+
+func (p KeyHashBasedShardProvider) ProvideShards(keys ...[]byte) (map[*DKVShard][][]byte, error) {
```
Comment (KalyanAkella) [inline clients/go/dkv/client.go:76]: Please include GoDoc for this public API

File: clients/go/dkv/client.go [hunk 3]
```diff
@@ -0,0 +1,248 @@
...
+	}
+	return nil, fmt.Errorf("valid DKV node type must be given")
+}
+
+type ShardProvider interface {
+	ProvideShard(key []byte) (*DKVShard, error)
+	ProvideShards(keys ...[]byte) (map[*DKVShard][][]byte, error)
+}
+
+type KeyHashBasedShardProvider struct {
```
Comment (KalyanAkella) [inline clients/go/dkv/client.go:57]: Please include GoDoc for this public struct

File: clients/go/dkv/client.go [hunk 4]
```diff
@@ -0,0 +1,248 @@
...
+func (s DKVShard) GetNodesByType(nodeType ...DKVServerRole) (*DKVNodeSet, error) {
+	for _, v := range nodeType {
+		if val, ok := s.Topology[v]; ok {
+			return val, nil
+		}
+	}
+	return nil, fmt.Errorf("valid DKV node type must be given")
+}
+
+type ShardProvider interface {
```
Comment (KalyanAkella) [inline clients/go/dkv/client.go:52]: Please include GoDoc

File: clients/go/dkv/client.go [hunk 5]
```diff
@@ -0,0 +1,248 @@
...
+	Name  string    `json:"name"`
+	Nodes []DKVNode `json:"nodes"`
+}
+
+type DKVShard struct {
+	Name     string                        `json:"name"`
+	Topology map[DKVServerRole]*DKVNodeSet `json:"topology"`
+}
+
+func (s DKVShard) GetNodesByType(nodeType ...DKVServerRole) (*DKVNodeSet, error) {
```
Comment (KalyanAkella) [inline clients/go/dkv/client.go:43]: Please include GoDoc

File: clients/go/dkv/client.go [hunk 6]
```diff
@@ -0,0 +1,248 @@
...
+	Host string `json:"host"`
+	Port int    `json:"port"`
+}
+
+type DKVNodeSet struct {
+	Name  string    `json:"name"`
+	Nodes []DKVNode `json:"nodes"`
+}
+
+type DKVShard struct {
```
Comment (KalyanAkella) [inline clients/go/dkv/client.go:38]: Please include GoDoc

File: clients/go/dkv/client.go [hunk 7]
```diff
@@ -0,0 +1,248 @@
...
+	}
+	return slaveRole
+}
+
+type DKVNode struct {
+	Host string `json:"host"`
+	Port int    `json:"port"`
+}
+
+type DKVNodeSet struct {
```
Comment (KalyanAkella) [inline clients/go/dkv/client.go:33]: Please include GoDoc

File: clients/go/dkv/client.go [hunk 8]
```diff
@@ -0,0 +1,248 @@
...
+)
+
+func GetNodeTypeByReadConsistency(rc serverpb.ReadConsistency) DKVServerRole {
+	if rc == serverpb.ReadConsistency_LINEARIZABLE {
+		return masterRole
+	}
+	return slaveRole
+}
+
+type DKVNode struct {
```
Comment (KalyanAkella) [inline clients/go/dkv/client.go:28]: Please include GoDoc

File: clients/go/dkv/client.go [hunk 9]
```diff
@@ -0,0 +1,248 @@
...
+
+type DKVServerRole string
+
+const (
+	noRole     DKVServerRole = "UNKNOWN"
+	masterRole               = "MASTER"
+	slaveRole                = "SLAVE"
+)
+
+func GetNodeTypeByReadConsistency(rc serverpb.ReadConsistency) DKVServerRole {
```
Comment (KalyanAkella) [inline clients/go/dkv/client.go:21]: Please include GoDoc. Else, check if this needs to be exposed.

File: clients/go/dkv/client.go [hunk 10]
```diff
@@ -0,0 +1,248 @@
...
+	"fmt"
+	"github.com/OneOfOne/xxhash"
+	"github.com/dgraph-io/ristretto"
+	"github.com/flipkart-incubator/dkv/pkg/ctl"
+	"github.com/flipkart-incubator/dkv/pkg/serverpb"
+
+	"log"
+)
+
+type DKVServerRole string
```
Comment (KalyanAkella) [inline clients/go/dkv/client.go:13]: Please include GoDoc. Else, check if this needs to be exposed.

File: clients/go/dkv/client.go [hunk 11]
```diff
@@ -0,0 +1,248 @@
...
+	return nil, fmt.Errorf("valid DKV node type must be given")
+}
+
+type ShardProvider interface {
+	ProvideShard(key []byte) (*DKVShard, error)
+	ProvideShards(keys ...[]byte) (map[*DKVShard][][]byte, error)
+}
+
+type KeyHashBasedShardProvider struct {
+	ShardConfiguration []DKVShard
```
Comment (KalyanAkella) [inline clients/go/dkv/client.go:58]: Check if this field needs to be exposed. If yes, include GoDoc.

File: clients/go/dkv/client.go [hunk 12]
```diff
@@ -0,0 +1,248 @@
...
+		if val, ok := s.Topology[v]; ok {
+			return val, nil
+		}
+	}
+	return nil, fmt.Errorf("valid DKV node type must be given")
+}
+
+type ShardProvider interface {
+	ProvideShard(key []byte) (*DKVShard, error)
+	ProvideShards(keys ...[]byte) (map[*DKVShard][][]byte, error)
```
Comment (KalyanAkella) [inline clients/go/dkv/client.go:54]: Check if the return value can be made more readable with a custom type. `[][]byte` might be hard to program for. Perhaps `KeyGrouping` or something similar ? We can apply similar changes on the Java side too, btw.

File: clients/go/dkv/client.go (hunk 11 above)
Comment (KalyanAkella) [inline clients/go/dkv/client.go:58]: Check if we should be using `[]*DKVShard` instead, as the type for this field.

File: clients/go/dkv/client.go [hunk 13]
```diff
@@ -0,0 +1,248 @@
...
+		shardId := p.getShardID(key)
+		shard := &p.ShardConfiguration[shardId]
+		m[shard] = append(m[shard], key)
+	}
+	return m, nil
+}
+
+type SimpleDKVClient struct {
+	*ctl.DKVClient
+	Addr string
```
Comment (KalyanAkella) [inline clients/go/dkv/client.go:88]: Please include GoDoc is this field is needed and should be exposed.

File: clients/go/dkv/client.go [hunk 14]
```diff
@@ -0,0 +1,248 @@
...
+}
+
+type ShardedDKVClient struct {
+	pool          *ristretto.Cache
+	shardProvider ShardProvider
+	shards        []DKVShard
```
Comment (KalyanAkella) [inline clients/go/dkv/client.go:94]: Check if we should be using `[]*DKVShard` instead, as the type for this field. Also, check if this field is really needed.

File: clients/go/dkv/client.go [hunk 15]
```diff
@@ -0,0 +1,248 @@
...
+	}
+	return m, nil
+}
+
+type SimpleDKVClient struct {
+	*ctl.DKVClient
+	Addr string
+}
+
+type ShardedDKVClient struct {
```
Comment (KalyanAkella) [inline clients/go/dkv/client.go:91]: Please include GoDoc.

File: clients/go/dkv/client.go [hunk 16]
```diff
@@ -0,0 +1,248 @@
...
+	pool          *ristretto.Cache
+	shardProvider ShardProvider
+	shards        []DKVShard
+}
+
+func NewShardedDKVClient(shardProvider ShardProvider) (*ShardedDKVClient, error) {
```
Comment (KalyanAkella) [inline clients/go/dkv/client.go:97]: Please include GoDoc.

Comment (KalyanAkella) [inline clients/go/dkv/client.go:97]: Please validate if the `shardProvider` is actually given.

File: clients/go/dkv/client.go [hunk 17]
```diff
@@ -0,0 +1,248 @@
...
+	}, nil
+}
+
+func (dkvClnt *ShardedDKVClient) getShardNode(shard *DKVShard, role ...DKVServerRole) (*DKVNodeSet, error) {
+	var nodeSet *DKVNodeSet
+	var err error
+	for _, serverRole := range role {
+		nodeSet, err = shard.GetNodesByType(serverRole)
+		if err == nil {
+			return nodeSet, err
```
Comment (KalyanAkella) [inline clients/go/dkv/client.go:124]: Please use `return nil, err`

**Review by KalyanAkella — CHANGES_REQUESTED** [review]

File: clients/go/dkv/client.go (hunk 17 above)
Comment (kingster) [reply clients/go/dkv/client.go:124]: This is an early return, breaks out of the iteration by returning `nodeSet`

File: clients/go/dkv/client.go [hunk 18]
```diff
@@ -0,0 +1,248 @@
...
+	m := make(map[*DKVShard][][]byte)
+	for _, key := range keys {
+		shardId := p.getShardID(key)
+		shard := &p.ShardConfiguration[shardId]
+		m[shard] = append(m[shard], key)
+	}
+	return m, nil
+}
+
+type SimpleDKVClient struct {
```
[the remark this answers is not in the document]
Comment (kingster) [reply clients/go/dkv/client.go]: Made this private. There is already an existing api to create `ctl.NewInSecureDKVClient(....)`

File: clients/go/dkv/client.go (hunk 1 above)
Comment (kingster) [reply clients/go/dkv/client.go]: Done

Comment (kingster) [reply clients/go/dkv/client.go]: Done

File: clients/go/dkv/client.go (hunk 2 above)
Comment (kingster) [reply clients/go/dkv/client.go]: Done

File: clients/go/dkv/client.go (hunk 3 above)
Comment (kingster) [reply clients/go/dkv/client.go:65]: Done

File: clients/go/dkv/client.go (hunk 4 above)
Comment (kingster) [reply clients/go/dkv/client.go:59]: Done

File: clients/go/dkv/client.go (hunk 5 above)
Comment (kingster) [reply clients/go/dkv/client.go]: Made it private

File: clients/go/dkv/client.go (hunk 6 above)
Comment (kingster) [reply clients/go/dkv/client.go:42]: Done

File: clients/go/dkv/client.go (hunk 7 above)
Comment (kingster) [reply clients/go/dkv/client.go:36]: Done

File: clients/go/dkv/client.go (hunk 8 above)
Comment (kingster) [reply clients/go/dkv/client.go:30]: Done

File: clients/go/dkv/client.go (hunk 9 above)
Comment (kingster) [reply clients/go/dkv/client.go]: Made private

File: clients/go/dkv/client.go (hunk 10 above)
Comment (kingster) [reply clients/go/dkv/client.go:14]: Done

File: clients/go/dkv/client.go (hunk 11 above)
Comment (kingster) [reply clients/go/dkv/client.go]: Made private.

File: clients/go/dkv/client.go (hunk 12 above)
Comment (kingster) [reply clients/go/dkv/client.go]: Done

File: clients/go/dkv/client.go (hunk 13 above)
Comment (kingster) [reply clients/go/dkv/client.go]: Made private.

File: clients/go/dkv/client.go (hunk 15 above)
Comment (kingster) [reply clients/go/dkv/client.go:106]: Done

File: clients/go/dkv/client.go (hunk 16 above)
Comment (kingster) [reply clients/go/dkv/client.go:112]: Done

File: clients/go/dkv/client.go (hunk 14 above)
Comment (kingster) [reply clients/go/dkv/client.go]: removed `shards` itself. This is not required as this is now available via `shardProvider`

File: clients/go/dkv/client.go (hunk 16 above)
Comment (kingster) [reply clients/go/dkv/client.go:112]: Done

File: clients/go/dkv/client.go (hunk 12 above)
[the remark this answers is not in the document]
Comment (kingster) [reply clients/go/dkv/client.go]: Refactored it to return slice of `DKVKeyGroup`

**Pull request merged**